app/auth/jupyterhub_auth.py

```diff
@@ -16,41 +16,40 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 """Implement JupyterHub authentication workflow."""
-
-import json
 import re
-from urllib.parse import parse_qs, urlencode, urljoin
+from urllib.parse import urlencode, urljoin
 
-import jwt
 import requests
-from oic import rndstr
 from flask import Blueprint, current_app, redirect, request, Response, session, url_for
 
-from .web import JWT_ALGORITHM, get_key_for_user
+from .oauth_provider_app import JupyterHubProviderApp
+from .utils import (
+    get_redis_key_from_session,
+    get_redis_key_from_token,
+    handle_login_request,
+    handle_token_request,
+)
 
+JH_SUFFIX = "jh_oauth_client"
 
 blueprint = Blueprint("jupyterhub_auth", __name__, url_prefix="/auth/jupyterhub")
 
 
 class JupyterhubUserToken:
     def process(self, request, headers):
-        from .. import store
 
         m = re.search(
             r"bearer (?P<token>.+)", headers.get("Authorization", ""), re.IGNORECASE
         )
         if m:
             # current_app.logger.debug('Authorization header present, token exchange')
             access_token = m.group("token")
-            decodentoken = jwt.decode(
-                access_token,
-                current_app.config["OIDC_PUBLIC_KEY"],
-                algorithms=JWT_ALGORITHM,
-                audience=current_app.config["OIDC_CLIENT_ID"],
+            jupyterhub_oauth_client = current_app.store.get_oauth_client(
+                get_redis_key_from_token(access_token, key_suffix=JH_SUFFIX)
+            )
+            headers["Authorization"] = "token {}".format(
+                jupyterhub_oauth_client.access_token
             )
-
-            jh_token = store.get(get_key_for_user(decodentoken, "jh_access_token"))
-            headers["Authorization"] = "token {}".format(jh_token.decode())
 
             # current_app.logger.debug('outgoing headers: {}'.format(json.dumps(headers)))
```

Comment:
Delete?

Reply:
done

Reply:
I think logging headers (though redacted) could be useful when debugging. I'd like to keep this around until https://github.com/SwissDataScienceCenter/renku-gateway/issues/113.